model/src/lib.rs

```diff
@@ -20,6 +22,29 @@ pub struct Signature {
     pub v: u8,
 }
 
+pub trait EIP712Signing {
+    fn digest(&self) -> [u8; 32];
+    fn signature(&self) -> Signature;
+    fn update_signature(&mut self, new_signature: Signature);
+
+    fn sign_self_with(&mut self, domain_separator: &DomainSeparator, key: &SecretKeyRef) {
+        let message = self
+            .signature()
+            .signing_digest_message(domain_separator, &self.digest());
+        // Unwrap because the only error is for invalid messages which we don't create.
+        let signature = Key::sign(key, &message, None).unwrap();
+        self.update_signature(Signature {
```

Comment:
Instead of adding `update_signature` on the trait, can this method instead return a `Signature` that you then set on self when you call it in the implementation of `default`?

Reply:
Sure I could do that but it is also used elsewhere (but only in the unit tests I believe).

Reply:
I suspect now that we don't even necessarily need to test `self_sign_with` anymore for both Cancellation and Creation structs since they are now redundant.

Reply:
All fixed. Looks much better now.